File: baekjoon/bj10866.cpp
/*
문제:백준10866번
출처:https://www.acmicpc.net/problem/10866
*/
#include<iostream>
#include<deque>
using namespace std;

int main() {
	deque<int>dq;
	string s;
	int N;
	cin >> N;
	while (N--) {
		int X;
		cin >> s;
		if (s == "push_front") {
			cin >> X;
			dq.push_front(X);
		}
		else if (s == "push_back") {
			cin >> X;
			dq.push_back(X);
		}
		else if (s == "pop_front") {
			if (!dq.empty()) {
				cout << dq.front() << '\n';
				dq.pop_front();
			}
			else cout << -1 << '\n';
		}
		else if (s == "pop_back") {
			if (!dq.empty()) {
				cout << dq.back() << '\n';
				dq.pop_back();
			}
			else cout << -1 << '\n';
		}
		else if (s == "size") {
			cout << dq.size() << '\n';
		}
		else if (s == "empty") {
			if (dq.empty())cout << 1 << '\n';
			else cout << 0 << '\n';
		}
		else if (s == "front") {
			if (!dq.empty())cout << dq.front() << '\n';
			else cout << -1 << '\n';
		}
		else {
			if (!dq.empty())cout << dq.back() << '\n';
			else cout << -1 << '\n';
		}
	}
	
}

File: 바킹독실전알고리즘/0x07/bj10866.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class bj10866 {
    static final int MX = 10005;
    static int[] dat = new int[MX*2+1];
    static int head = MX, tail=MX;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        StringBuilder sb = new StringBuilder();
        for(int i=0;i<N;i++){
            String command = br.readLine();
            String[] commands = command.split(" ");
            if(commands[0].equals("push_front")){
                push_front(Integer.parseInt(commands[1]));
            }else if(commands[0].equals("push_back")){
                push_back(Integer.parseInt(commands[1]));
            }else if(commands[0].equals("pop_front")){
                if(head==tail)sb.append(-1).append("\n");
                else {
                    sb.append(front()).append("\n");
                    pop_front();
                }
            }else if(commands[0].equals("pop_back")){
                if(head==tail)sb.append(-1).append("\n");
                else {
                    sb.append(back()).append("\n");
                    pop_back();
                }
            }else if(commands[0].equals("size")){
                sb.append(size()).append("\n");
            }else if(commands[0].equals("empty")){
                sb.append(empty()).append("\n");
            }else if(commands[0].equals("front")){
                if(head==tail)sb.append(-1).append("\n");
                else {
                    sb.append(front()).append("\n");
                }
            }else if(commands[0].equals("back")){
                if(head==tail)sb.append(-1).append("\n");
                else {
                    sb.append(back()).append("\n");
                }
            }
        }
        System.out.println(sb);
    }

    private static int empty() {
        if(tail==head) return 1;
        else return 0;
    }

    private static int front() {
        return dat[head];
    }

    private static int back() {
        return dat[tail-1];
    }
    private static void pop_front() {
        head++;
    }
    private static void pop_back() {
        tail--;
    }

    private static int size() {
        return tail-head;
    }

    private static void push_front(int x) {
        dat[--head] = x;
    }
    private static void push_back(int x) {
        dat[tail++] = x;
    }
}
